src/memory/cgroup_v2.rs

```diff
@@ -0,0 +1,163 @@
+#[cfg(target_os = "linux")]
+pub mod os {
+    use std::io::BufRead;
+    use std::path::Path;
+
+    use super::super::Memory;
+    use crate::{file_to_buf_reader, parse_u64};
+    use crate::{path_to_string, ProbeError, Result};
+
+    pub fn read_and_parse_v2_sys_memory(path: &Path) -> Result<Memory> {
+        let mut memory = Memory {
+            total: None,
+            free: None,
+            used: 0,
+            buffers: None,
+            cached: None,
+            shmem: None,
+            swap_total: None,
+            swap_free: None,
+            swap_used: None,
+        };
+
+        memory.total = read_file_value_as_u64(&path.join("memory.max"))
+            .ok()
+            .map(bytes_to_kilo_bytes);
+
+        memory.used = bytes_to_kilo_bytes(read_file_value_as_u64(&path.join("memory.current"))?);
+
+        let reader = file_to_buf_reader(&path.join("memory.stat"))?;
+        for line_result in reader.lines() {
+            let line = line_result.map_err(|e| ProbeError::IO(e, path_to_string(path)))?;
+            let segments: Vec<&str> = line.split_whitespace().collect();
+            let value = parse_u64(&segments[1])?;
+
+            if segments[0] == "shmem" {
+                memory.shmem = Some(bytes_to_kilo_bytes(value));
+                break;
+            };
+        }
+
+        memory.free = memory.total.map(|total| total - memory.used);
+
+        memory.swap_total = match read_file_value_as_u64(&path.join("memory.swap.max")) {
+            Ok(value) => memory
+                .total
+                .map(|total| bytes_to_kilo_bytes(value).saturating_sub(total)),
+            Err(_) => None,
+        };
+        memory.swap_used = read_file_value_as_u64(&path.join("memory.swap.current"))
+            .ok()
+            .map(bytes_to_kilo_bytes);
+        memory.swap_free = memory
+            .swap_total
+            .zip(memory.swap_used)
+            .map(|(total, used)| total.saturating_sub(used));
+
+        Ok(memory)
+    }
+
+    fn read_file_value_as_u64(path: &Path) -> Result<u64> {
+        let mut reader = file_to_buf_reader(path)?;
+        let mut line = String::new();
+        reader
+            .read_line(&mut line)
+            .map_err(|e| ProbeError::IO(e, path_to_string(path)))?;
+        parse_u64(&line.trim())
+    }
+
+    fn bytes_to_kilo_bytes(bytes: u64) -> u64 {
```

Comment:
You can move this function up to a shared module, like the memory or cgroup one, so it's not implemented twice.